diff -u netdata-1.10.0.orig/plugins.d/python.d.plugin netdata-1.10.0/plugins.d/python.d.plugin
--- netdata-1.10.0.orig/plugins.d/python.d.plugin	2018-05-08 20:01:40.000000000 +0300
+++ netdata-1.10.0/plugins.d/python.d.plugin	2018-05-08 20:06:57.000000000 +0300
@@ -15,10 +15,8 @@
 from sys import version_info, argv
 from time import sleep
 
-try:
-    from time import monotonic as time
-except ImportError:
-    from time import time
+#   from time import monotonic as time
+from time import time
 
 PY_VERSION = version_info[:2]
 PLUGIN_CONFIG_DIR = os.getenv('NETDATA_CONFIG_DIR', os.path.dirname(__file__) + '/../../../../etc/netdata') + '/'
@@ -32,10 +30,7 @@
 from bases.loggers import PythonDLogger
 from bases.collection import setdefault_values, run_and_exit
 
-try:
-    from collections import OrderedDict
-except ImportError:
-    from third_party.ordereddict import OrderedDict
+from ordereddict import OrderedDict
 
 BASE_CONFIG = {'update_every': os.getenv('NETDATA_UPDATE_EVERY', 1),
                'retries': 60,
