--- a/src/flb_network.c
+++ b/src/flb_network.c
@@ -523,9 +523,10 @@
             }
 
             /* Connection is broken, not much to do here */
-            str = strerror_r(error, so_error_buf, sizeof(so_error_buf));
+            /* XXX: XSI */
+            int _err = strerror_r(error, so_error_buf, sizeof(so_error_buf));
             flb_error("[net] TCP connection failed: %s:%i (%s)",
-                      u->tcp_host, u->tcp_port, str);
+                      u->tcp_host, u->tcp_port, so_error_buf);
             return -1;
         }
     }
